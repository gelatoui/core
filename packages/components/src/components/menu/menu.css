:host {
  display: block;
  box-sizing: border-box;
  position: relative;
  --glu-menu-min-height: 64px;
  --glu-menu-mobile-height: 56px;
  --glu-menu-background: var(--tokens-background-base-surface-1);
  --glu-menu-border-color: var(--tokens-stroke-on-light-default);
  --glu-menu-shadow: 0 2px 4px rgba(0, 0, 0, 0.05);
  --glu-menu-padding: var(--spacing-100) var(--spacing-300);
  --glu-menu-logo-max-height: 40px;
  --glu-menu-transition: all 0.3s ease;
  --glu-menu-mobile-breakpoint: 768px;
  --glu-menu-mobile-menu-bg: var(--tokens-background-base-surface-1);
  --glu-menu-toggle-color: var(--tokens-text-on-light-default);
  --glu-menu-z-index: 999;
  --glu-menu-logo-max-width: 200px;
}

:host(.glu-menu--sticky) {
  position: sticky;
  top: 0;
  z-index: var(--glu-menu-z-index);
}

.glu-menu__nav {
  box-sizing: border-box;
  width: 100%;
  min-height: var(--glu-menu-min-height);
  background-color: var(--glu-menu-background);
  border-bottom: 1px solid var(--tokens-stroke-on-light-default, #DCE0E6);
  display: flex;
  align-items: center;
}

/* Menu Container */
.glu-menu__container {
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 100%;
  padding: var(--glu-menu-padding);
  width: 100%;
  max-width: 1440px;
  margin: 0 auto;
  flex-wrap: wrap;
}

/* Base layout for sections */
.glu-menu__logo-wrapper,
.glu-menu__items-wrapper,
.glu-menu__actions-wrapper {
  display: flex;
  align-items: center;
}

.glu-menu__logo-wrapper ::slotted(*),
.glu-menu__items-wrapper ::slotted(*) {
  display: flex;
  align-items: center;
  gap: var(--spacing-100);
  display: flex;
  flex-wrap: wrap;
}

.glu-menu__items-wrapper glu-menu-item::part(menu-item) {
  padding: var(--spacing-100, 0.5rem) var(--spacing-200, 1rem);
}

.glu-menu__actions-wrapper ::slotted(*) {
  display: flex;
  align-items: center;
  gap: var(--spacing-150)
}

.glu-menu__logo-wrapper {
  flex-shrink: 0;
  display: flex;
  align-items: center;
  max-width: var(--glu-menu-logo-max-width);
  height: 100%;
}

.glu-menu__logo-wrapper img,
.glu-menu__logo-wrapper ::slotted(*)
{
  max-width: 100%;
  width: auto;
  height: auto;
  max-height: 100%;
  max-height: var(--glu-menu-logo-max-height);
  object-fit: contain;
  padding: var(--spacing-050) 0;
}

/* Menu items wrapper */
.glu-menu__items-wrapper {
  /* flex: 1; */
  justify-content: center;
}

/* Right actions wrapper */
.glu-menu__actions-wrapper {
  flex-shrink: 0;
}

/* Logo Left Layout */
.glu-menu__container--logo-left {
  justify-content: flex-start;
  gap: var(--spacing-300);
}

.glu-menu__container--logo-left .glu-menu__items-wrapper {
  justify-content: flex-start;
}

.glu-menu__container--logo-left .glu-menu__actions-wrapper {
  margin-left: auto;
}

/* Logo Center Layout */
.glu-menu__container--logo-center {
  justify-content: space-between;
}

.glu-menu__container--logo-center .glu-menu__items-wrapper {
  justify-content: flex-start;
}

/* Menu Center Layout */
.glu-menu__container--menu-center {
  justify-content: space-between;
  gap: var(--spacing-300);
}

/* Mobile styles */
:host(.glu-menu--mobile) .glu-menu__nav {
  height: var(--glu-menu-mobile-height);
  padding: 0 var(--spacing-200);
}

.glu-menu__mobile-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 100%;
}

.glu-menu__mobile-toggle {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 40px;
  height: 40px;
  background: transparent;
  border: none;
  cursor: pointer;
  padding: 0;
}

.glu-menu__mobile-toggle-icon,
.glu-menu__mobile-toggle-icon::before,
.glu-menu__mobile-toggle-icon::after {
  display: block;
  width: 24px;
  height: 2px;
  background-color: var(--glu-menu-toggle-color);
  position: relative;
  transition: var(--glu-menu-transition);
}

.glu-menu__mobile-toggle-icon::before,
.glu-menu__mobile-toggle-icon::after {
  content: '';
  position: absolute;
  left: 0;
}

.glu-menu__mobile-toggle-icon::before {
  top: -8px;
}

.glu-menu__mobile-toggle-icon::after {
  bottom: -8px;
}

:host(.glu-menu--mobile-open) .glu-menu__mobile-toggle-icon {
  background-color: transparent;
}

:host(.glu-menu--mobile-open) .glu-menu__mobile-toggle-icon::before {
  top: 0;
  transform: rotate(45deg);
}

:host(.glu-menu--mobile-open) .glu-menu__mobile-toggle-icon::after {
  bottom: 0;
  transform: rotate(-45deg);
}

.glu-menu__mobile-menu {
  display: none;
  position: absolute;
  top: var(--glu-menu-mobile-height);
  left: 0;
  width: 100%;
  background-color: var(--glu-menu-mobile-menu-bg);
  box-shadow: var(--glu-menu-shadow);
  z-index: var(--glu-menu-z-index);
  padding: var(--spacing-200);
  box-sizing: border-box;
  border-top: 1px solid var(--glu-menu-border-color);
  max-height: 0;
  overflow: hidden;
  transition: var(--glu-menu-transition);
}

:host(.glu-menu--mobile-open) .glu-menu__mobile-menu {
  display: block;
  max-height: calc(100vh - var(--glu-menu-mobile-height));
  overflow-y: auto;
}

.glu-menu__mobile-items {
  display: flex;
  flex-direction: column;
  padding-bottom: var(--spacing-200);
}

.glu-menu__mobile-items ::slotted(*) {
  display: flex;
  flex-direction: column;
  gap: var(--spacing-100);
}

.glu-menu__mobile-actions {
  display: flex;
  flex-direction: column;
  border-top: 1px solid var(--glu-menu-border-color);
  padding-top: var(--spacing-200);
}

/* Media Queries */
/* @media (max-width: 767px) {
  :host(:not(.glu-menu--mobile)) .glu-menu__container {
    flex-wrap: wrap;
  }

  :host(:not(.glu-menu--mobile)) .glu-menu__items-wrapper {
    order: 3;
    width: 100%;
    margin: 0;
    padding: var(--spacing-100) 0;
  }
} */

.glu-menu__item {
  font-family: var(--body-medium-medium-font-family);
  font-size: var(--body-medium-medium-font-size);
  font-weight: var(--body-medium-medium-font-weight);
  line-height: var(--body-medium-medium-line-height);
}

.glu-menu__item--active {
  font-weight: var(--body-medium-bold-font-weight);
  color: var(--tokens-primary-default);
}

.glu-menu__mobile-items .glu-menu__item {
  font-family: var(--body-medium-regular-font-family);
  font-size: var(--body-medium-regular-font-size);
  font-weight: var(--body-medium-regular-font-weight);
  line-height: var(--body-medium-regular-line-height);
  padding: var(--spacing-100) 0;
}
