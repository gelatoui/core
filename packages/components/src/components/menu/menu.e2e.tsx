/* eslint-disable no-undef */
import { newE2EPage } from '@stencil/core/testing'

describe('glu-menu e2e', () => {
  it('renders the default component', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-menu></glu-menu>')

    const element = await page.find('glu-menu')

    expect(element).not.toBeNull()

    expect(element).toHaveClass('glu-menu')
  })

  it('renders the logo prop correctly', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-menu logo="https://placehold.co/600x400"></glu-menu>')

    // Query the image inside the shadow DOM
    const img = await page.find('glu-menu >>> .glu-menu__logo-img')

    expect(img).not.toBeNull()

    expect(img.getAttribute('src')).toBe('https://placehold.co/600x400')
  })

  it('applies the sticky class when is-sticky is set', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-menu is-sticky></glu-menu>')

    const element = await page.find('glu-menu')

    expect(element).toHaveClass('glu-menu--sticky')
  })

  it('applies the mobile class when is-responsive is set and viewport is mobile', async () => {
    const page = await newE2EPage()

    // Set viewport to a mobile size (width < 768px)
    await page.setViewport({ width: 500, height: 800 })

    await page.setContent('<glu-menu is-responsive></glu-menu>')

    await page.waitForChanges()

    const element = await page.find('glu-menu')

    expect(element).toHaveClass('glu-menu--mobile')
  })
})
