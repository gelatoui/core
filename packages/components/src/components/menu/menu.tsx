import { Component, Element, forceUpdate, Fragment, h, Host, Listen, Prop, State, Watch } from '@stencil/core'

/**
 * @component
 * @tag glu-menu
 * @shadow true
 *
 * @slot logo - Slot for custom logo content (used when `logo` prop is not set).
 * @slot menu-items - Slot for the menu items to be displayed.
 * @slot right-actions - Slot for additional actions or buttons aligned to the right.
 */
@Component({
  tag: 'glu-menu',
  styleUrl: 'menu.css',
  shadow: true
})
export class GluMenu {
  /**
   * Defines the menu layout type.
   * - `logo-left`: Logo on the left, menu items in the center, right actions on the right.
   * - `logo-center`: Menu items on the left, logo in the center, right actions on the right.
   * - `menu-center`: Logo on the left, menu items in the center, right actions on the right.
   */
  @Prop({ reflect: true }) readonly type: 'logo-left' | 'logo-center' | 'menu-center' = 'logo-left'

  /**
   * URL for the logo image. If set, it takes precedence over the `logo` slot.
   */
  @Prop() readonly logo?: string

  /**
   * If `true`, the menu will be fixed at the top of the viewport.
   */
  @Prop() readonly isSticky: boolean = false

  /**
   * If `true`, the menu adapts for smaller screens by enabling mobile-friendly behavior.
   */
  @Prop() readonly isResponsive: boolean = false

  /**
   * Reference to the host element.
   */
  // eslint-disable-next-line no-undef
  @Element() hostElement!: HTMLGluMenuElement

  /**
   * Tracks whether the menu is currently in mobile view.
   */
  @State() isMobileView = false

  /**
   * Tracks whether the mobile menu is open.
   */
  @State() isMobileMenuOpen = false

  @Listen('resize', { target: 'window' })
  onResize() {
    forceUpdate(this)
  }

  componentWillLoad() {
    this.checkViewportSize()
  }

  @Watch('isResponsive')
  responsiveChanged() {
    if (this.isResponsive) {
      this.checkViewportSize()
    } else {
      this.isMobileView = false
    }
  }

  /**
   * Checks whether the viewport width falls within the mobile threshold.
   * @local This function sets `isMobileView` to true if the window width is less than 768px and if the menu is responsive.
   */
  private checkViewportSize() {
    const width = window.innerWidth

    this.isMobileView = this.isResponsive && width < 768
  }

  /**
   * Toggles the mobile menu open/close state.
   */
  private toggleMobileMenu = () => {
    this.isMobileMenuOpen = !this.isMobileMenuOpen
  }

  /**
   * Renders the logo either from the `logo` prop or the `logo` slot.
   */
  private renderLogo() {
    return this.logo ?
      (
        <img src={this.logo} alt="Logo" class="glu-menu__logo-img" />
      ) :
      (
        <slot name="logo"></slot>
      )
  }

  /**
   * Renders the mobile menu toggle button.
   */
  private renderMobileToggle() {
    return (
      <button
        class="glu-menu__mobile-toggle"
        onClick={this.toggleMobileMenu}
        aria-expanded={this.isMobileMenuOpen ? 'true' : 'false'}
        aria-label="Toggle menu"
      >
        <span class="glu-menu__mobile-toggle-icon"></span>
      </button>
    )
  }

  /**
   * Renders the `logo-left` layout.
   */
  private renderLogoLeft() {
    return (
      <div class="glu-menu__container glu-menu__container--logo-left">
        <div class="glu-menu__logo-wrapper">{this.renderLogo()}</div>
        <div class="glu-menu__items-wrapper"><slot name="menu-items"></slot></div>
        <div class="glu-menu__actions-wrapper"><slot name="right-actions"></slot></div>
      </div>
    )
  }

  /**
   * Renders the `logo-center` layout.
   */
  private renderLogoCenter() {
    return (
      <div class="glu-menu__container glu-menu__container--logo-center">
        <div class="glu-menu__items-wrapper left"><slot name="menu-items"></slot></div>
        <div class="glu-menu__logo-wrapper">{this.renderLogo()}</div>
        <div class="glu-menu__actions-wrapper right"><slot name="right-actions"></slot></div>
      </div>
    )
  }

  /**
   * Renders the `menu-center` layout.
   */
  private renderMenuCenter() {
    return (
      <div class="glu-menu__container glu-menu__container--menu-center">
        <div class="glu-menu__logo-wrapper left">{this.renderLogo()}</div>
        <div class="glu-menu__items-wrapper"><slot name="menu-items"></slot></div>
        <div class="glu-menu__actions-wrapper right"><slot name="right-actions"></slot></div>
      </div>
    )
  }

  /**
   * Renders the mobile menu panel.
   */
  private renderMobileMenu() {
    return (
      <div
        class={{
          'glu-menu__mobile-menu': true,
          'glu-menu__mobile-menu--open': this.isMobileMenuOpen
        }}
      >
        <div class="glu-menu__mobile-items">
          <slot name="menu-items"></slot>
        </div>
        <div class="glu-menu__mobile-actions">
          <slot name="right-actions"></slot>
        </div>
      </div>
    )
  }

  render() {
    this.checkViewportSize()

    return (
      <Host
        class={{
          'glu-menu': true,
          [`glu-menu--${this.type}`]: true,
          'glu-menu--sticky': this.isSticky,
          'glu-menu--mobile': this.isMobileView,
          'glu-menu--mobile-open': this.isMobileView && this.isMobileMenuOpen
        }}
      >
        <nav class="glu-menu__nav" role="navigation">
          {this.isMobileView ?
            (
              <div class="glu-menu__mobile-header">
                <div class="glu-menu__logo-wrapper">{this.renderLogo()}</div>
                {this.renderMobileToggle()}
              </div>
            ) :
            (
              <Fragment>
                {this.type === 'logo-left' && this.renderLogoLeft()}
                {this.type === 'logo-center' && this.renderLogoCenter()}
                {this.type === 'menu-center' && this.renderMenuCenter()}
              </Fragment>
            )}
          {this.isMobileView && this.renderMobileMenu()}
        </nav>
      </Host>
    )
  }
}
