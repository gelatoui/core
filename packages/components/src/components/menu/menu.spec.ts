/* eslint-disable no-undef */
import { GluMenu } from './menu'

import { newSpecPage } from '@stencil/core/testing'

const ResizeObserverMock = jest.fn(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn()
}))

global.ResizeObserver = ResizeObserverMock

describe('glu-menu', () => {
  it('renders correctly with default props', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu></glu-menu>'
    })

    expect(page.root).toMatchSnapshot()

    expect(page.root).toHaveClass('glu-menu')
  })

  it('renders correctly with logo prop', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu logo="logo.png"></glu-menu>'
    })

    expect(page.root).toMatchSnapshot()

    expect(page.root.shadowRoot.querySelector('img')).toHaveAttribute('src')
  })

  it('renders slots correctly', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: `
        <glu-menu>
          <span slot="logo">Logo Slot</span>
          <span slot="menu-items">Menu Items</span>
          <span slot="right-actions">Right Actions</span>
        </glu-menu>`
    })

    expect(page.root).toMatchSnapshot()

    expect(page.root.querySelector('[slot="logo"]').textContent).toBe('Logo Slot')

    expect(page.root.querySelector('[slot="menu-items"]').textContent).toBe('Menu Items')

    expect(page.root.querySelector('[slot="right-actions"]').textContent).toBe('Right Actions')
  })

  it('applies isSticky class when prop is set', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu is-sticky></glu-menu>'
    })

    expect(page.root).toHaveClass('glu-menu--sticky')
  })

  it('sets mobile view when isResponsive and width < 768px', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu is-responsive></glu-menu>'
    })

    Object.defineProperty(page.win, 'innerWidth', {
      configurable: true,
      value: 500
    })

    page.win.dispatchEvent(new Event('resize'))

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--mobile')
  })

  it('toggles mobile menu open/close on button click', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu is-responsive></glu-menu>'
    })

    Object.defineProperty(page.win, 'innerWidth', {
      configurable: true,
      value: 500
    })

    page.win.dispatchEvent(new Event('resize'))

    await page.waitForChanges()

    const toggleButton = page.root.shadowRoot.querySelector('.glu-menu__mobile-toggle') as HTMLElement

    expect(toggleButton).toBeTruthy()

    // First click to open
    toggleButton.click()

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--mobile-open')

    // Second click to close
    toggleButton.click()

    await page.waitForChanges()

    expect(page.root).not.toHaveClass('glu-menu--mobile-open')
  })

  it('renders logo-center layout when type is set to logo-center', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu type="logo-center"></glu-menu>'
    })

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--logo-center')

    expect(page.root.shadowRoot.querySelector('.glu-menu__container--logo-center')).toBeTruthy()
  })

  it('renders menu-center layout when type is set to menu-center', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu type="menu-center"></glu-menu>'
    })

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--menu-center')

    expect(page.root.shadowRoot.querySelector('.glu-menu__container--menu-center')).toBeTruthy()
  })

  it('uses slot for logo when logo prop is not set', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: `
      <glu-menu>
        <span slot="logo">Logo Slot</span>
      </glu-menu>
    `
    })

    await page.waitForChanges()

    expect(page.root.querySelector('[slot="logo"]').textContent).toBe('Logo Slot')
  })

  it('updates mobile view on window resize', async () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1000 })

    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu is-responsive></glu-menu>'
    })

    await page.waitForChanges()

    expect(page.root).not.toHaveClass('glu-menu--mobile')

    // Simulate window resize
    window.innerWidth = 500

    window.dispatchEvent(new Event('resize'))

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--mobile')
  })

  it('applies sticky and mobile classes when both isSticky and isResponsive are set', async () => {
    const page = await newSpecPage({
      components: [GluMenu],
      html: '<glu-menu is-sticky is-responsive></glu-menu>'
    })

    Object.defineProperty(page.win, 'innerWidth', {
      configurable: true,
      value: 500
    })

    page.win.dispatchEvent(new Event('resize'))

    await page.waitForChanges()

    expect(page.root).toHaveClass('glu-menu--sticky')

    expect(page.root).toHaveClass('glu-menu--mobile')
  })
})
