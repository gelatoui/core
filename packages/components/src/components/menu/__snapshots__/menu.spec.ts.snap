// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`glu-menu renders correctly with default props 1`] = `
<glu-menu class="glu-menu glu-menu--logo-left" type="logo-left">
  <template shadowrootmode="open">
    <nav class="glu-menu__nav" role="navigation">
      <div class="glu-menu__container glu-menu__container--logo-left">
        <div class="glu-menu__logo-wrapper">
          <slot name="logo"></slot>
        </div>
        <div class="glu-menu__items-wrapper">
          <slot name="menu-items"></slot>
        </div>
        <div class="glu-menu__actions-wrapper">
          <slot name="right-actions"></slot>
        </div>
      </div>
    </nav>
  </template>
</glu-menu>
`;

exports[`glu-menu renders correctly with logo prop 1`] = `
<glu-menu class="glu-menu glu-menu--logo-left" logo="logo.png" type="logo-left">
  <template shadowrootmode="open">
    <nav class="glu-menu__nav" role="navigation">
      <div class="glu-menu__container glu-menu__container--logo-left">
        <div class="glu-menu__logo-wrapper">
          <img alt="Logo" class="glu-menu__logo-img" src="logo.png">
        </div>
        <div class="glu-menu__items-wrapper">
          <slot name="menu-items"></slot>
        </div>
        <div class="glu-menu__actions-wrapper">
          <slot name="right-actions"></slot>
        </div>
      </div>
    </nav>
  </template>
</glu-menu>
`;

exports[`glu-menu renders slots correctly 1`] = `
<glu-menu class="glu-menu glu-menu--logo-left" type="logo-left">
  <template shadowrootmode="open">
    <nav class="glu-menu__nav" role="navigation">
      <div class="glu-menu__container glu-menu__container--logo-left">
        <div class="glu-menu__logo-wrapper">
          <slot name="logo"></slot>
        </div>
        <div class="glu-menu__items-wrapper">
          <slot name="menu-items"></slot>
        </div>
        <div class="glu-menu__actions-wrapper">
          <slot name="right-actions"></slot>
        </div>
      </div>
    </nav>
  </template>
  <span slot="logo">
    Logo Slot
  </span>
  <span slot="menu-items">
    Menu Items
  </span>
  <span slot="right-actions">
    Right Actions
  </span>
</glu-menu>
`;
